libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/StreamingExtensions/BotFrameworkWebSocketAdapter.cs

```diff
@@ -0,0 +1,94 @@
+    {
+        private readonly IChannelProvider _channelProvider;
+        private readonly ICredentialProvider _credentialProvider;
+        
+
+        public BotFrameworkWebSocketAdapter(ICredentialProvider credentialProvider, IChannelProvider channelProvider = null, ILogger<BotFrameworkWebSocketAdapter> logger = null)
+        {
+            this._credentialProvider = credentialProvider;
+            this._channelProvider = channelProvider;
+        }
+
+        public ConcurrentDictionary<string, WebSocketServer> Connections { get; set; }
+
+        public async Task ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken = default(CancellationToken))
+         {
+            if (httpRequest == null)
+            {
+                throw new ArgumentNullException(nameof(httpRequest));
+            }
+
+            if (httpResponse == null)
+            {
+                throw new ArgumentNullException(nameof(httpResponse));
+            }
+
+            if (bot == null)
+            {
+                throw new ArgumentNullException(nameof(bot));
+            }
+
+            var authHeader = httpRequest.Headers["Authorization"];
+            var channelId = httpRequest.Headers["ChannelId"];
+            try
+            {
+                var claimsIdentity = await JwtTokenValidation.ValidateAuthHeader(authHeader, _credentialProvider, _channelProvider, channelId).ConfigureAwait(false);
+                if (!claimsIdentity.IsAuthenticated)
+                {
+                    httpRequest.HttpContext.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+                    return;
+                }
+            }
+            catch (Exception)
+            {
+                httpRequest.HttpContext.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+                return;
+            }
+
+            if (!httpRequest.HttpContext.WebSockets.IsWebSocketRequest)
+            {
+                httpRequest.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+                await httpRequest.HttpContext.Response.WriteAsync("Upgrade to WebSocket required.").ConfigureAwait(false);
+                return;
+            }
+
+            await CreateWebSocketConnectionAsync(httpRequest.HttpContext, authHeader, channelId, bot).ConfigureAwait(false);
+        }
+
+        public async Task CreateWebSocketConnectionAsync(HttpContext httpContext, string authHeader, string channelId, IBot bot)
+        {
+            var socket = await httpContext.WebSockets.AcceptWebSocketAsync().ConfigureAwait(false);
+            var serverID = Guid.NewGuid().ToString();
+            var server = new WebSocketServer(socket, new StreamingExtensionRequestHandler(new BotFrameworkStreamingExtensionsAdapter(serverID), bot));
+
+            try
+            {
+                WebSocketServerRegistry.RegisterNewServer(serverID, server);
```

Comment:
Notes for later:
1. We need to clean up this dictionary periodically (for closed sockets).
2. need to decide what to do if you get multiple from the same host...this is probably just fine.